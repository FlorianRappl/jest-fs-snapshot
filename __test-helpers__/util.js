import path from 'path';

export const getFixture = name => path.resolve(process.cwd(), '__tests__/__fixtures__', name);

export const getAsar = name => path.resolve(process.cwd(), '__tests__/__fs_snapshots__', `${name}.asar`);
