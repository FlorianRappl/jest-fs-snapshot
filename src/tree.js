import path from 'path';
import fs from 'fs';

export function getDirectoryTree(src, options = {}) {
    const stats = fs.lstatSync(src);

    if (options.callback) {
        options.callback(path.resolve(src), stats);
    }

    if (stats.isDirectory()) {
        const files = fs.readdirSync(src).reduce((res, child) => {
            if (!(options.exclude && options.exclude.test(child))) {
                res[child] = getDirectoryTree(path.resolve(src, child), options);
            }
            return res;
        }, {});

        return {files};
    }

    if (stats.isFile()) {
        return {
            size: stats.size,
        };
    }

    return {};
}

function getNodeType(node, pathToNode) {
    if (typeof node === 'object') {
        if (typeof node.files === 'object') return 'directory';
        if (typeof node.size === 'number') return 'file';
        if (typeof node.link === 'string') return 'symlink';
    }
    throw new Error(`Unknown node type '${JSON.stringify(node)}' at ${pathToNode}.`);
}

const ROOT = './';

export function compareDirectoryTrees({actual, expected, pathToNode = ROOT, onFile}) {
    const actualType = getNodeType(actual, pathToNode);
    const expectedType = getNodeType(expected, pathToNode);

    if (pathToNode === ROOT && (actualType !== 'directory' || expectedType !== 'directory')) {
        throw new Error(`The roots must be directories but got a ${actualType} and a ${expectedType}.`);
    }

    if (actualType !== expectedType) {
        throw new Error(`Nodes at '${pathToNode}' have different types: a ${actualType} and a ${expectedType}.`);
    }

    if (actualType === 'file') {
        if (onFile) {
            onFile({actual, expected, pathToNode});
            return null;
        }

        if (actual.size !== expected.size) {
            throw new Error(`Files at '${pathToNode}' have different size: ${actual.size} and ${expected.size}.`);
        }
    }

    if (actualType === 'directory') {
        const actualKeys = Object.keys(actual.files);
        const expectedKeys = Object.keys(expected.files);

        for (let i = 0, l = actualKeys.length; i < l; i++) {
            const key = actualKeys[i];
            const bKeyIndex = expectedKeys.indexOf(key);
            if (bKeyIndex === -1) {
                throw new Error(
                    `Directories at ${pathToNode} are different: expected directory doesn't have a node '${key}'.`
                );
            }
            expectedKeys.splice(bKeyIndex, 1);
        }

        if (expectedKeys.length > 0) {
            throw new Error(
                `Directories at ${pathToNode} are different: actual directory doesn't have node(s) '${expectedKeys.join(
                    "', '"
                )}'.`
            );
        }

        actualKeys.forEach(key => {
            compareDirectoryTrees({
                actual: actual.files[key],
                expected: expected.files[key],
                pathToNode: pathToNode + (pathToNode !== ROOT ? '/' : '') + key,
                onFile,
            });
        });
    }

    return null;
}
