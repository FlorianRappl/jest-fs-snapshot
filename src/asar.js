import asar from 'asar';
import asarDisk from 'asar/lib/disk';
import {getDirectoryTree} from './tree';

export function getAsarTree(archive) {
    return asarDisk.readArchiveHeaderSync(archive).header;
}

export function createAsarPackage(src, dest, {exclude, ...options} = {}) {
    const files = [];
    const metadata = {};
    function callback(path, stat) {
        let type;
        if (stat.isDirectory()) {
            type = 'directory';
        }
        if (stat.isFile()) {
            type = 'file';
        }
        if (stat.isSymbolicLink()) {
            type = 'symbolic';
        }

        if (!type) {
            throw new Error('ataatt');
        }

        files.push(path);
        metadata[path] = {type, stat};
    }

    getDirectoryTree(src, {exclude, callback});

    return asar.createPackageFromFiles(src, dest, files, metadata, options, err => (err ? reject(err) : resolve()));
}
