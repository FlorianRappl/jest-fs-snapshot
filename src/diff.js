import path from 'path';
import fs from 'fs';
import isValidUTF8 from 'utf-8-validate';
import iconv from 'iconv-lite';
import diff from 'jest-diff';
import {NO_DIFF_MESSAGE} from 'jest-diff/build/constants';
import {EXPECTED_COLOR, RECEIVED_COLOR} from 'jest-matcher-utils';
import asarDisk from 'asar/lib/disk';
import {getDirectoryTree, compareDirectoryTrees} from './tree';
import {getAsarTree} from './asar';

export function getBufferEncoding(buffer) {
    const size = buffer.length;
    if (size === 0) {
        return 'utf8';
    }

    // UTF-8 BOM
    if (size >= 3 && buffer[0] === 0xef && buffer[1] === 0xbb && buffer[2] === 0xbf) {
        return 'utf8';
    }

    // UTF-16 BE BOM
    if (size >= 2 && buffer[0] === 0xfe && buffer[1] === 0xff) {
        return 'utf16-be';
    }

    // UTF-16 LE BOM
    if (size >= 2 && buffer[0] === 0xff && buffer[1] === 0xfe) {
        return 'utf16';
    }

    if (isValidUTF8(buffer)) {
        return 'utf8';
    }

    return 'binary';
}

export function compareBuffers(actualBuf, expectedBuf) {
    const actualEncoding = getBufferEncoding(actualBuf);
    const expectedEncoding = getBufferEncoding(expectedBuf);

    if (actualEncoding !== expectedEncoding) {
        return (
            ` Comparing files with different encodings. Expected ${EXPECTED_COLOR(expectedEncoding)}` +
            ` but received ${RECEIVED_COLOR(actualEncoding)}.`
        );
    }

    if (actualEncoding === 'binary') {
        return Buffer.compare(actualBuf, expectedBuf) === 0 ? NO_DIFF_MESSAGE : 'Comparing different binary files.';
    }

    const actualText = iconv.decode(actualBuf, actualEncoding);
    const expectedText = iconv.decode(expectedBuf, expectedEncoding);

    return diff(actualText, expectedText);
}

export function compareDirectoryToAsar(actualPath, expectedPath, options) {
    try {
        const actualTree = getDirectoryTree(actualPath, options);
        const expectedTree = getAsarTree(expectedPath);

        const asarFilesystem = asarDisk.readFilesystemSync(expectedPath);

        compareDirectoryTrees({
            actual: actualTree,
            expected: expectedTree,
            onFile({pathToNode}) {
                if (path.sep === '\\') {
                    // eslint-disable-next-line no-param-reassign
                    pathToNode = pathToNode.replace(/[/]/g, '\\');
                }

                const thisPath = path.resolve(actualPath, pathToNode);
                const actual = fs.readFileSync(thisPath);
                const expected = asarDisk.readFileSync(asarFilesystem, pathToNode, asarFilesystem.getFile(pathToNode));

                let compareFunc = compareBuffers;

                if (options.customCompare) {
                    const match = options.customCompare.filter(({check}) => check(thisPath))[0];
                    if (match) compareFunc = match.compare || compareFunc;
                }

                const res = compareFunc(actual, expected);

                if (res !== NO_DIFF_MESSAGE) {
                    const error = new Error(res);
                    error.pathToNode = pathToNode;
                    throw error;
                }
            },
        });
    } catch (error) {
        return error;
    }

    return null;
}

export {NO_DIFF_MESSAGE};
