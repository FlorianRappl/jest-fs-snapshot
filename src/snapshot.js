import fs from 'fs';
import path from 'path';
import filenamify from 'filenamify';
import {matcherHint, EXPECTED_COLOR, RECEIVED_COLOR} from 'jest-matcher-utils';
import {createAsarPackage} from './asar';
import {compareDirectoryToAsar} from './diff';

// eslint-disable-next-line import/prefer-default-export
export function matchFilesystemSnapshot({received, expected, context, name}) {
    const {testPath, currentTestName, isNot, snapshotState} = context;

    if (isNot) {
        throw new Error('Jest: `.not` cannot be used with `.toMatchFilesystemSnapshot()`.');
    }

    if (!snapshotState) {
        throw new Error('Jest: snapshot state must be initialized.');
    }

    const typeOfReceived = typeof received;
    if (typeOfReceived !== 'string') {
        throw new Error(`Jest: expected to receive a path to a directory but got ${typeOfReceived}.`);
    }

    const count = Number(snapshotState._counters.get(currentTestName) || 0);
    const testKey = `${path.basename(testPath)} ${currentTestName} ${count}`;
    const snapshotFilename = `${filenamify(testKey)}.snap.asar`;
    const snapshotsDir = path.resolve(testPath, '..', '__fs_snapshots__');

    const typeOfExpected = typeof expected;
    let expectedFullPath = expected;
    if (typeOfExpected === 'undefined') {
        expectedFullPath = path.resolve(snapshotsDir, snapshotFilename);
    } else if (typeOfExpected !== 'string') {
        throw new Error(`Jest: a matcher expected a path to a snapshot but got ${typeOfExpected}.`);
    }

    const receivedFullPath = path.resolve(received);
    if (!fs.existsSync(receivedFullPath)) {
        throw new Error(`Jest: expected path \`${received}\` doesn't exists.`);
    }

    const hasSnapshot = fs.existsSync(expectedFullPath);

    if (!hasSnapshot) {
        if (snapshotState._updateSnapshot === 'none') {
            return {
                name,
                pass: false,
                report: () =>
                    `New snapshot was ${RECEIVED_COLOR('not written')}. The update flag ` +
                    `must be explicitly passed to write a new snapshot.\n\n` +
                    `This is likely because this test is run in a continuous integration ` +
                    `(CI) environment in which snapshots are not written by default.`,
            };
        }

        createAsarPackage(receivedFullPath, expectedFullPath, {exclude: /node_modules/});

        snapshotState.added += 1;

        return {
            name,
            pass: true,
            message: () => '',
        };
    }

    const diffMessage = compareDirectoryToAsar(receivedFullPath, expectedFullPath, {
        exclude: /node_modules/,
        aAnnotation: 'Snapshot',
        bAnnotation: 'Received',
        expand: snapshotState.expand,
    });

    if (!diffMessage) {
        snapshotState.matched += 1;
        snapshotState._uncheckedKeys.delete(snapshotFilename);

        return {
            name,
            pass: true,
            message: () => '',
        };
    }

    if (snapshotState._updateSnapshot === 'all') {
        createAsarPackage(receivedFullPath, expectedFullPath, {exclude: /node_modules/});

        snapshotState.updated += 1;
        snapshotState._uncheckedKeys.delete(snapshotFilename);

        return {
            name,
            pass: true,
            message: () => '',
        };
    }

    snapshotState.unmatched += 1;
    snapshotState._uncheckedKeys.delete(snapshotFilename);

    const report = () =>
        `${RECEIVED_COLOR('Received value')} does not match ${EXPECTED_COLOR(
            `stored snapshot${diffMessage.pathToNode ? ` at ${diffMessage.pathToNode}` : ''}`
        )}.\n\n ${diffMessage.message}`;

    return {
        name,
        pass: false,
        message: () => `${matcherHint(`.${name}`, 'value', '')}\n\n${report()}`,
        report,
    };
}
