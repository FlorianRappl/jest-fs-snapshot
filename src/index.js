import {matchFilesystemSnapshot} from './snapshot';

// eslint-disable-next-line import/prefer-default-export
export function toMatchFilesystemSnapshot(received, expected) {
    return matchFilesystemSnapshot({received, expected, context: this, name: 'toMatchFilesystemSnapshot'});
}
