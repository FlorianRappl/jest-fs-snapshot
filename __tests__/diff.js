import path from 'path';
import fs from 'fs-extra';
import {getBufferEncoding, compareBuffers, NO_DIFF_MESSAGE} from '../src/diff';
import {getFixture} from '../__test-helpers__/util';

describe('getBufferEncoding()', () => {
    const encodingsFixtures = getFixture('encodings');

    it('should decode utf8.txt', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'utf8.txt'));
        expect(getBufferEncoding(buf)).toEqual('utf8');
    });

    it('should decode utf8_bom.txt', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'utf8_bom.txt'));
        expect(getBufferEncoding(buf)).toEqual('utf8');
    });

    it('should decode utf8_emojis.txt', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'utf8_emojis.txt'));
        expect(getBufferEncoding(buf)).toEqual('utf8');
    });

    it('should decode utf8_shishi.txt', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'utf8_shishi.txt'));
        expect(getBufferEncoding(buf)).toEqual('utf8');
    });

    it('should decode utf8_not-cp949.txt', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'utf8_not-cp949.txt'));
        expect(getBufferEncoding(buf)).toEqual('utf8');
    });

    it('should decode utf16-be.txt', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'utf16-be.txt'));
        expect(getBufferEncoding(buf)).toEqual('utf16-be');
    });

    it('should decode utf16-be_bom.txt', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'utf16-be_bom.txt'));

        expect(getBufferEncoding(buf)).toEqual('utf16-be');
    });

    it('should decode utf16-le_bom.txt', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'utf16-le_bom.txt'));
        expect(getBufferEncoding(buf)).toEqual('utf16');
    });

    it('should detect ansi.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'ansi.txt'));
        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect big5.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'big5.txt'));
        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect big5_B.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'big5_B.txt'));
        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect cp936_with-bad-chars.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'cp936_with-bad-chars.txt'));
        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect euc_jp.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'euc_jp.txt'));

        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect euc_tw.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'euc_tw.txt'));
        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect euc_tw_B.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'euc_tw_B.txt'));
        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect gb.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'gb.txt'));
        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect gb2.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'gb2.txt'));

        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect kr.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'kr.txt'));
        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect latin.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'latin.txt'));
        expect(getBufferEncoding(buf)).toEqual('binary');
    });

    it('should detect sjis.txt as binary', async () => {
        const buf = await fs.readFile(path.join(encodingsFixtures, 'sjis.txt'));
        expect(getBufferEncoding(buf)).toEqual('binary');
    });
});

describe('compareBuffers()', () => {
    const encodingsFixtures = getFixture('encodings');
    const diffFixtures = getFixture('diff');

    it('should return an error if binary files are different', async () => {
        const buf1 = await fs.readFile(path.join(encodingsFixtures, 'ansi.txt'));
        const buf2 = await fs.readFile(path.join(encodingsFixtures, 'big5.txt'));
        expect(compareBuffers(buf1, buf2)).toMatchInlineSnapshot(`"Comparing different binary files."`);
    });

    it('should return ok if binary files are the same', async () => {
        const buf1 = await fs.readFile(path.join(encodingsFixtures, 'ansi.txt'));
        const buf2 = await fs.readFile(path.join(encodingsFixtures, 'ansi.txt'));
        expect(compareBuffers(buf1, buf2)).toBe(NO_DIFF_MESSAGE);
    });

    it('should return an error if encodings are different', async () => {
        const buf1 = await fs.readFile(path.join(encodingsFixtures, 'ansi.txt'));
        const buf2 = await fs.readFile(path.join(encodingsFixtures, 'utf8.txt'));
        expect(compareBuffers(buf1, buf2)).toMatchInlineSnapshot(
            `" Comparing files with different encodings. Expected [32mutf8[39m but received [31mbinary[39m."`
        );

        const buf3 = await fs.readFile(path.join(encodingsFixtures, 'utf8.txt'));
        const buf4 = await fs.readFile(path.join(encodingsFixtures, 'utf16-be.txt'));
        expect(compareBuffers(buf3, buf4)).toMatchInlineSnapshot(
            `" Comparing files with different encodings. Expected [32mutf16-be[39m but received [31mutf8[39m."`
        );
    });

    it('should return no changes for the same file', async () => {
        const buf1 = await fs.readFile(path.join(diffFixtures, 'lorem-ipsum-1.txt'));
        const buf2 = await fs.readFile(path.join(diffFixtures, 'lorem-ipsum-1.txt'));
        expect(compareBuffers(buf1, buf2)).toBe(NO_DIFF_MESSAGE);
    });

    it('should return a diff for lorem-ipsum files', async () => {
        const buf1 = await fs.readFile(path.join(diffFixtures, 'lorem-ipsum-1.txt'));
        const buf2 = await fs.readFile(path.join(diffFixtures, 'lorem-ipsum-2.txt'));
        expect(compareBuffers(buf1, buf2)).toMatchSnapshot();
    });

    it('should return a diff for emoji files', async () => {
        const buf1 = await fs.readFile(path.join(diffFixtures, 'emoji-1.txt'));
        const buf2 = await fs.readFile(path.join(diffFixtures, 'emoji-2.txt'));
        expect(compareBuffers(buf1, buf2)).toMatchSnapshot();
    });
});
