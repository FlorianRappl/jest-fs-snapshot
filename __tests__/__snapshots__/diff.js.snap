// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compareBuffers() should return a diff for emoji files 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[32m- This is a text file ðŸŒˆ[39m
[31m+ This is a text file too ðŸ¦„[39m
[2m  [22m"
`;

exports[`compareBuffers() should return a diff for lorem-ipsum files 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[2m  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.[22m
[2m  Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.[22m
[32m- Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.[39m
[2m  Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.[22m
[2m  [22m"
`;
